### PR TITLE
feat(ckpt): checkpoint bug fixes and feature enhancements.

### PR DESCRIPTION
**Pull request by SolenoidWGT**
## Motivation

1. - [x] add unittests for `CheckpointManager`.
2. - [x] fix `StorageManager` get_fns bug.
3. - [x] fix `CheckpointManager` querying latest boto3/local checkpoint bug when snapshot not found.
4. - [x] fix `CheckpointManager`  returned latest local checkpoint path not be set 'local:' prefix.
5. - [x] validate checkpoint can be stored normally at the very beginning of training.
6. - [x] optimize the logic of load ckpt and remove redundancy argument.
7. - [x] fix random initialization zero fp32 buffer overwrite pre-loaded model weights when only_model_load.

## Modification

* internlm/utils/model_checkpoint.py
* internlm/utils/storage_manager.py
* internlm/core/trainer.py
* train.py

## BC-breaking (Optional)

1. ` config.ckpt.load_ckpt_folder` now support dict input, format like:
```python
# where the ’path‘ indicate ckpt path, 
# the 'content‘ means which components will be loaded, now support input is "model", "sampler", "optimizer",
# the ’ckpt_type‘ means the type of checkpoint to be loaded, now only 'normal' type is supported.

load_ckpt_folder= dict(path=MODEL_ONLY_FOLDER, content=["model", "sampler", "optimizer"], ckpt_type="normal"),
```

2. change `config.ckpt.load_given_ckpt` to `config.ckpt.auto_resume_latest_ckpt`(load_given_ckpt is wtill working, set `load_given_ckpt=True` is equal with `auto_resume_latest_ckpt=False`)
3. remove `config.ckpt.load_optimizer`
4. remove `config.ckpt.load_model_only_folder`

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: configs/7B_sft.py [hunk 1]
```diff
@@ -22,13 +22,10 @@
 ckpt = dict(
     enable_save_ckpt=False,  # enable ckpt save.
     save_ckpt_folder=SAVE_CKPT_FOLDER,  # Path to save training ckpt.
-    # load_ckpt_folder=LOAD_CKPT_FOLDER, # Ckpt path to resume training(load weights and scheduler/context states).
-    # load_model_only_folder=MODEL_ONLY_FOLDER, # Path to initialize with given model weights.
-    load_optimizer=True,  # Wheter to load optimizer states when continuing training.
+    load_ckpt_folder= dict(path=MODEL_ONLY_FOLDER, content=["model"], ckpt_type="normal"),
```
Comment (yhcc) [inline configs/7B_sft.py:25]: 加入一个新的load_ckpt_kwargs来承接dict这种配置

Comment (yhcc) [reply configs/7B_sft.py:25]: content不要用list，而用tuple

Comment (yhcc) [reply configs/7B_sft.py:25]: 要么加一个说明，比如content："all", "model", "optimizer", "sampler", "scheduler"

File: internlm/core/trainer.py [hunk 2]
```diff
@@ -61,26 +71,22 @@ def __str__(self) -> str:
...
 
-        # track the actual updates of sampler when using weighted sampling
-        if hasattr(self, "batch_sampler"):
-            self.batch_sampler = train_dl.batch_sampler.copy()
-            self.batch_sampler_iter = iter(self.batch_sampler)
+        # Because the ckpt save occurs after updating 'step_count',
+        # there is no need to increment 'step_count' here (Does our step count start from 0 ?),
+        # However, 'batch_count' is updating before ckpt storage, so it need to inc 1 when resume.
+        self.batch_count = other_stuffs["batch_count"] + 1  # here you need to shift a batch backward
+        self.step_count = other_stuffs.get("step_count", other_stuffs["batch_count"])
```
Comment (yhcc) [inline internlm/core/trainer.py:89]: self.step_count = other_stuffs.get("step_count", other_stuffs["batch_count"]+1)

File: internlm/utils/tests/common_fixture.py
Comment (yhcc) [inline internlm/utils/tests/common_fixture.py:1]: 这个移动到internlm/tests/utils/common_fixture.py

File: internlm/utils/model_checkpoint.py [hunk 3]
```diff
@@ -32,6 +33,47 @@
 class CheckpointType(Enum):
     NORMAL_CHECKPOINT = 1
     SNAPSHOT_CHECKPOINT = 2
+    ONLY_MODEL_CHECKPOINT = 3
```
Comment (yhcc) [inline internlm/utils/model_checkpoint.py:36]: 这个需要删掉

File: configs/7B_sft.py (hunk 1 above)
Comment (SolenoidWGT) [reply configs/7B_sft.py]: Done

File: internlm/train/training_internlm.py [hunk 4]
```diff
@@ -410,13 +410,7 @@ def record_current_batch_training_metrics(
...
                 panel_metrics[norm_key] = norm_value
 
-            logger.info(
-                "{line}",
-                line=line,
-                extra=panel_metrics,
-            )
-        else:
-            logger.info(line)
+        logger.info(line)
```
Comment (gaoyang07) [inline internlm/train/training_internlm.py:413]: `panel_metrics` missed?

File: internlm/utils/model_checkpoint.py [hunk 5]
```diff
@@ -283,19 +316,37 @@ def __init__(self, ckpt_config, model, model_config=None, model_config_file=None
...
-        self.checkpoint_every = ckpt_config.checkpoint_every
-        self.save_ckpt_folder = ckpt_config.save_ckpt_folder
-        self.snapshot_ckpt_folder = ckpt_config.snapshot_ckpt_folder
-        self.oss_snapshot_freq: int = ckpt_config.oss_snapshot_freq
-        self.stop_file_path = ckpt_config.stop_file_path
-        self.load_model_only_folder = ckpt_config.load_model_only_folder
+        self.enable_save_ckpt = get_config_value(ckpt_config, "enable_save_ckpt", False)
+        self.checkpoint_every = get_config_value(ckpt_config, "checkpoint_every", 1)
+        self.save_ckpt_folder = get_config_value(ckpt_config, "save_ckpt_folder", None)
+        self.oss_snapshot_freq: int = get_config_value(ckpt_config, "oss_snapshot_freq", 1)
```
Comment (gaoyang07) [inline internlm/utils/model_checkpoint.py:322]: Set the default value of `self.oss_snapshot_freq` to a larger num, e.g., 50. So does the default value of `self.checkpoint_every` above.

File: internlm/utils/model_checkpoint.py [hunk 6]
```diff
@@ -306,22 +357,73 @@ def __init__(self, ckpt_config, model, model_config=None, model_config_file=None
...
+                    if load_ctx.lower() == "scheduler":
+                        load_mask |= CheckpointLoadMask.LOAD_SCHEDULER
+                    if load_ctx.lower() == "sampler":
+                        load_mask |= CheckpointLoadMask.LOAD_SAMPLER
+                    if load_ctx.lower() == "all":
+                        load_mask |= CheckpointLoadMask.LOAD_ALL
+
+                self.load_ckpt_info["content"] = load_mask
+
+            assert set([self.load_ckpt_info["ckpt_type"]]).issubset(set(["normal"]))
```
Comment (gaoyang07) [inline internlm/utils/model_checkpoint.py:419]: here.

**Review by gaoyang07 — APPROVED** [review]

File: train.py [hunk 7]
```diff
@@ -35,7 +35,8 @@
     parse_args,
 )
 from internlm.utils.evaluation import evaluate_on_val_dls
-from internlm.utils.gputest import bench_gpu, bench_net
+
+# from internlm.utils.gputest import bench_gpu, bench_net
```
Comment (gaoyang07) [inline train.py:39]: delete this

**Review by gaoyang07 — APPROVED** [review]

**Pull request merged**